src/tasks/task_list.hpp

```diff
@@ -46,28 +47,102 @@ class TaskList {
   void ClearComplete() {
     auto task = task_list_.begin();
     while (task != task_list_.end()) {
-      if (task->IsComplete()) {
+      if (task->GetStatus() == TaskStatus::complete &&
+          task->GetType() != TaskType::iterative) {
         task = task_list_.erase(task);
       } else {
         ++task;
       }
     }
   }
+  void IterationComplete(const std::string &label) {
+    completed_iters_.insert(label);
+    for (auto &task : task_list_) {
+      if (task.GetLabel() == label) {
+        task.SetStatus(TaskStatus::complete);
+      }
+    }
+  }
+  void ClearIteration(const std::string &label) {
+    auto task = task_list_.begin();
+    while (task != task_list_.end()) {
+      if (task->GetLabel() == label) {
+        task = task_list_.erase(task);
+      } else {
+        ++task;
+      }
+    }
+  }
+  void ResetIteration(const std::string &label) {
+    count_[label]++;
+    if (count_[label] == max_iterations_[label]) {
+      if (throw_with_max_iters_[label]) {
+        PARTHENON_THROW("Iteration " + label +
+                        " reached maximum allowed cycles without convergence.");
+      }
+      if (warn_with_max_iters_[label]) {
+        PARTHENON_WARN("Iteration " + label +
+                       " reached maximum allowed cycles without convergence.");
+      }
+      IterationComplete(label);
+      return;
+    }
+    for (auto &task : task_list_) {
+      if (task.GetLabel() == label) {
+        if (tasks_completed_.CheckDependencies(task.GetID())) {
+          tasks_completed_.SetFinished(task.GetID());
+        }
+        task.SetStatus(TaskStatus::incomplete);
+      }
+    }
+  }
   TaskListStatus DoAvailable() {
     for (auto &task : task_list_) {
+      // first skip task if it's complete.  Only possible for TaskType::iterative
+      if (task.GetStatus() == TaskStatus::complete) continue;
       auto dep = task.GetDependency();
       if (tasks_completed_.CheckDependencies(dep)) {
-        TaskStatus status = task();
-        if (status == TaskStatus::complete) {
-          task.SetComplete();
+        task();
+        if (task.GetStatus() == TaskStatus::complete) {
           MarkTaskComplete(task.GetID());
+          if (task.GetType() == TaskType::completion_criteria) {
+            IterationComplete(task.GetLabel());
+          }
+        } else if (task.GetStatus() == TaskStatus::iterate &&
+                   task.GetType() == TaskType::completion_criteria) {
+          ResetIteration(task.GetLabel());
         }
       }
     }
```

Comment:
This is beginning to get a little hard to follow, but I'm not sure there's anything to be done about that.